core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2219,6 +2219,14 @@ protected Scope visitSetOperation(SetOperation node, Optional<Scope> scope)
                         oldField.getOriginTable(),
                         oldField.getOriginColumnName(),
                         oldField.isAliased());
+
+                int index = i; // Variable used in Lambda should be final
+                analysis.addSourceColumns(
+                        outputDescriptorFields[index],
+                        childrenTypes.stream()
```

Comment:
For `INTERSECT`, and particularly for `EXCEPT`, we could report only the source columns from the leftmost relation.
I'm not sure of the intended semantics, though.

Reply:
Well, it does not matter much. We care about columns being used in the operation. 

Also for security reasons if one column is restricted, and you know other columns then you could "calculate" the restricted column data. Notice the case `A EXCEPT B` where B is restricted and A contains the whole domain. Given that I would prefer that anything that is created from ``A EXCEPT B` would be also restricted.

Notice one could say that `coalesce(a, b, c)` is hitting the same issue when `a` is not nullable, and we might have more corner cases like this.